### ISSUE TITLE
Trivy Operator crashes large clusters

### ISSUE THREAD

**Issue by bo0ts** [aquasecurity/trivy-operator#2335]
**What steps did you take and what happened:**

We ran trivy operator installed through the Helm chart version `0.25.0` on an OKD 4.15 cluster with ~1200 non-system pods, most of them running with two containers or more. We expected this to create quite some load on the cluster and customized the installation this way:

```
excludeNamespaces: "kube-system,openshift-*"
operator:
  vulnerabilityScannerEnabled: true
  sbomGenerationEnabled: true
  # disable config, rbac, infra and compliance scanner
  configAuditScannerEnabled: false
  rbacAssessmentScannerEnabled: false
  infraAssessmentScannerEnabled: false
  clusterComplianceEnabled: false
```

**What did you expect to happen:**

We expected this to take quite a while to complete and initially results looked good. At some point load on etcd got exceedingly high and leader changes started to occur. This led to a complete crash of the API and the cluster had to be restored from a recent etcd backups. We managed to take a snapshot before starting recovery and the etcd cluster grew from ~380MB to ~1.2Gi

We would need to scale the control-plane significantly to accommodate trivy-operator and also consider deploying it in a production cluster to be too dangerous. 

Are there any options to reduce the footprint of CRs? In general I feel that trivy-operator relies too much on etcd as an actual data storage and that isn't in line with kubernetes design philosophy.

**Comment by afdesk** [comment]
@bo0ts thanks for the report!
we should investigate it.
Did this issue happen before or it was introduced in Chart v0.25.0?

**Comment by itaysk** [comment]
If I understand correctly, this is expected by Trivy Operator's design which relies exclusively on CRDs (etcd) to store scans output. If that is the issue then it isn't a bug, but enhancement request, and I'm afraid we are not planning to change the fundementals of trivy operator at this point in time.

**Comment by bo0ts** [comment]
> Did this issue happen before or it was introduced in Chart v0.25.0?

This was the first version we installed so I do not know if this is a regression.



> If I understand correctly, this is expected by Trivy Operator's design which relies exclusively on CRDs (etcd) to store scans output. If that is the issue then it isn't a bug, but enhancement request, and I'm afraid we are not planning to change the fundementals of trivy operator at this point in time.

That's a pity. What's the alternative? Using trivy with custom scripting?

**Comment by itaysk** [comment]
> That's a pity. What's the alternative? Using trivy with custom scripting?

Allow me to ellaborate: Trivy Operator's purpose is to be a [Kubernetes Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) for Trivy. Kubernetes Operators' design is centered around Kubernetes Resources (CRDs). Kubernetes Resources have limitations imposed by Kubernetes, etcd, or your cluster configuration. That's part of the deal of being a Kubernetes native operator.
If there are suggestions for incrementally improving Trivy Operator within it's current design and scope (Kubernetes Operator, CRDs), than we can consider (it would help if it's a PR).

Alternatively if you don't need/want a Kubernetes Operator but just a solution for scanning kubernetes workloads, you can roll your own orchastration for Trivy like you suggested, or consider a commercial offering that is [built for scale such as Aqua Security](https://github.com/aquasecurity/resources/blob/main/trivy-aqua.md#kubernetes-scanning)

**Comment by bo0ts** [comment]
> Trivy. Kubernetes Operators' design is centered around Kubernetes Resources (CRDs). Kubernetes Resources have limitations imposed by Kubernetes, etcd, or your cluster configuration. That's part of the deal of being a Kubernetes native operator.

I think that is stretching the concept of an operator, but this is expected to happen given that it is such a flexible concept.

Another (and IMO more common)  way to think about operators is to represent concepts of the application as custom resource definitions and instances of those concepts as custom resources. The operators task then is to reflect custom resources in the application and reflect application state back into the status of the custom resource. Consider the rabbitmq-topology operator where there are CRDs for RabbitMQ objects like a RabbitMQ instance, User, Queue, or VHost. The operator then makes sure custom resources have an associated representation in the "real" world. No one would expect the custom resource to hold any queue messages or other application state.

The current design effectively (ab)uses the control-plane as a data-store for trivy (and maybe out of necessity because trivy has no built-in data-plane). I'd rather have an actual data-store for trivy and would let the operator handle the scheduling of scans (You could think of custom resources like a TrivyStore, Scan, ScanSchedule, ClusterScan, ClusterScanSchedule,...).

I assume you would consider that out of scope of the current design?

**Comment by simar7** [comment]
> Did this issue happen before or it was introduced in Chart v0.25.0? 

@bo0ts could you give another version a try? We're curious if the latest version changed anything significant.

**Comment by afdesk** [comment]
> > Did this issue happen before or it was introduced in Chart v0.25.0?
> 
> @bo0ts could you give another version a try? We're curious if the latest version changed anything significant.

@simar7 it doesn't look like the issue depends on the latest version. there are a few similar issues about etcd problem.

**Comment by itaysk** [comment]
> I'd rather have an actual data-store for trivy and would let the operator handle the scheduling of scans ... I assume you would consider that out of scope of the current design?

yes, this is currently out of scope for Trivy Operator open source project

**Comment by bo0ts** [comment]
> yes, this is currently out of scope for Trivy Operator open source project

Too bad, but understandable. Let's close this.
